contrib/smtp_proxy/filters/network/source/smtp_filter.cc

```diff
@@ -0,0 +1,346 @@
+#include "contrib/smtp_proxy/filters/network/source/smtp_filter.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/connection.h"
+
+#include "source/common/common/assert.h"
+#include "source/extensions/filters/network/well_known_names.h"
+
+#include "contrib/smtp_proxy/filters/network/source/smtp_decoder.h"
+
+using ConfigProto = envoy::extensions::filters::network::smtp_proxy::v3alpha::SmtpProxy;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+SmtpFilterConfig::SmtpFilterConfig(const SmtpFilterConfigOptions& config_options,
+                                   Stats::Scope& scope)
+    : downstream_ssl_(config_options.downstream_ssl_), upstream_ssl_(config_options.upstream_ssl_),
+      scope_{scope}, stats_{generateStats(config_options.stats_prefix_, scope)} {}
+
+SmtpFilter::SmtpFilter(SmtpFilterConfigSharedPtr config) : config_{config} {
+  if (!decoder_) {
+    decoder_ = createDecoder();
+  }
+}
+
+Network::FilterStatus SmtpFilter::onNewConnection() {
+  ENVOY_CONN_LOG(trace, "smtp_proxy: onNewConnection()", read_callbacks_->connection());
+
+  config_->stats_.sessions_.inc();
+
+  state_ = UpstreamGreeting;
+
+  return Network::FilterStatus::Continue;
+}
+
+// Network::ReadFilter
+Network::FilterStatus SmtpFilter::onData(Buffer::Instance& data, bool) {
+  ENVOY_CONN_LOG(trace, "smtp_proxy: onData {} bytes, state {}", read_callbacks_->connection(),
+                 data.length(), state_);
+
+  switch (state_) {
+  case Passthrough:
+    return Network::FilterStatus::Continue;
+  case DownstreamEhlo:
+  case DownstreamStarttls:
+    break;
+  default:
+    // client spoke out of turn -> hang up
+    config_->stats_.sessions_downstream_desync_.inc();
+    read_callbacks_->connection().close(Network::ConnectionCloseType::NoFlush);
+    return Network::FilterStatus::StopIteration;
+  }
```

Comment:
The initial value of `state_` is `UpstreamGreeting`, then when the first data arriving, the connection will be closed directly?
Or I missed something?

Reply:
ahh... in SMTP the server/upstream sends first so we expect onWrite() here first, if the client sends first (onRead()), that's a protocol error